### PR TITLE
Fix node placement collision detection failure [The Ghost Node]

### LINKED ISSUE #9625: Block placing allowed in occupied space if in mid-air
##### Minetest version
```
5.2.0
```

##### OS / Hardware
Operating system: Linux Mint 19.3 64-bit
CPU: Intel Core i5-4690K

##### Summary
You can place a block while partially occupying the target area, being in mid-air.

##### Steps to reproduce
Jump and, before you land, place a block under you. You will end up inside the block you just placed, without even taking damage. "Bonus:" your character will turn completely black.

screenshot_20200410_092347

### PR DESCRIPTION
**Pull request by jordan4ibanez**
Add compact, short information about your PR for easier understanding:

- Goal of the PR
You know when you place a node and you can fall into the node or just straight up build upon yourself even when that's not supposed to happen? That's what this fixes. Or in technical terms:
Fixes node placement prediction collision detection.
- How does the PR work?
We throw out the old code where it uses point based and we swap it for collision based.
- Does it resolve any reported issue?
No idea
- Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
I sure hope so.

This PR is a Ready for Review.

If you would like to see the final product in action: https://youtu.be/trBhd3vxybM?feature=shared

EDIT by grorp: Fixes #9625

## How to test

Try to phase into a node while placing it.

That's about it really. This has been a bug for 10+ years and it's not going to be a bug anymore.

### CODE REVIEW

File: src/client/game.cpp [hunk 1]
```diff
@@ -3731,14 +3732,37 @@ bool Game::nodePlacement(const ItemDefinition &selected_def,
 
 		// Don't place node when player would be inside new node
 		// NOTE: This is to be eventually implemented by a mod as client-side Lua
```
Comment (grorp) [inline src/client/game.cpp:3734]: You can probably remove this comment, I doubt this is ever going to happen.

File: src/client/game.cpp [hunk 2]
```diff
@@ -3731,14 +3732,37 @@ bool Game::nodePlacement(const ItemDefinition &selected_def,
...
+		// We are simply colliding the player's current collisionbox in the world to a
+		// "normal node" collision box. This can be improved to take nodeboxes.
+		bool narrow_phase_collision = false;
+		if (!skip_check) {
+
+			// Create a "normal node" collision box.
+			aabb3f normal_node_cbox(
+				(f32)p.X - 0.5f, (f32)p.Y - 0.5f, (f32)p.Z - 0.5f,
+				(f32)p.X + 0.5f, (f32)p.Y + 0.5f, (f32)p.Z + 0.5f
+			);
```
Comment (grorp) [inline src/client/game.cpp:3748]: This should use the actual collision box(es) of the predicted node (`predicted_node.getCollisionBoxes`).

File: src/client/game.cpp [hunk 3]
```diff
@@ -3731,14 +3732,37 @@ bool Game::nodePlacement(const ItemDefinition &selected_def,
...
+
+			// Create the player's collision box.
+			aabb3f cbox = player->getCollisionbox();
+			v3f pos = player->getPosition() / v3f{10.0f, 10.0f, 10.0f};			
+			const auto min = (cbox.MinEdge / v3f{10.0f, 10.0f, 10.0f}) + pos;
+			const auto max = (cbox.MaxEdge / v3f{10.0f, 10.0f, 10.0f}) + pos;
```
Comment (grorp) [inline src/client/game.cpp:3754]: You should use `BS` instead of hardcoded `10.0f`.

Even better, you can avoid dividing by BS entirely by doing the collision in BS-space. The only thing that has to be converted to BS-space for that is the node position `p`. Everything else you need is already in BS-space.

**Review by grorp — CHANGES_REQUESTED** [review]
While this PR seems to be correct, it makes it much harder than before to place nodes below yourself while jumping in MTG. I don't know what can be done about that, but I think something needs to be done about it.

File: src/client/game.cpp (hunk 1 above)
Comment (jordan4ibanez) [reply src/client/game.cpp:3734]: sounds like a plan

File: src/client/game.cpp (hunk 2 above)
Comment (jordan4ibanez) [reply src/client/game.cpp:3747]: Let's take a look and see what I can do about this. Keep in mind: This now turns into an iteration! Nodeboxes and all that.

**Comment by jordan4ibanez** [comment]
> While this PR seems to be correct, it makes it much harder than before to place nodes below yourself while jumping in MTG. I don't know what can be done about that, but I think something needs to be done about it.

Give 2be4be1a2c92ddbcd39b5c9c0938e452c9c692a5 a try :D

**Comment by jordan4ibanez** [comment]
I have no idea why this is failing on visual studio so if any core dev knows why please inform me

File: src/client/game.cpp [hunk 4]
```diff
@@ -3730,15 +3731,56 @@ bool Game::nodePlacement(const ItemDefinition &selected_def,
...
+					rawCbox.MaxEdge + BS_space_p
+				);
+
+				// Now collide it. Narrow phase is done.
+				narrow_phase_collision = player_cbox.intersectsWithBox(normal_node_cbox);
+				if (narrow_phase_collision)
+					break;
+			}
+			// Release the memory.
+			free(cboxes);
```
Comment (v-rob) [inline src/client/game.cpp:3771]: Ack--BAD!  Don't use `free()` on something allocated with `new`!  That's major undefined behavior.

Better yet, do `std::vector<aabb3f> cboxes;` above if possible; avoid heap allocation whenever possible, especially raw allocation with `new`.

File: src/client/game.cpp [hunk 5]
```diff
@@ -3730,15 +3731,56 @@ bool Game::nodePlacement(const ItemDefinition &selected_def,
...
-					neighborpos != player->getStandingNodePos() + v3s16(0, 2, 0))) {
+				(client->checkPrivilege("noclip") && g_settings->getBool("noclip"));
+
+		// We are simply colliding the player's current collisionbox in the world to a
+		// "normal node" collision box. This can be improved to take nodeboxes.
+		bool narrow_phase_collision = false;
+		if (!skip_check) {
+			// Collect all node collision boxes.
+			auto cboxes = new std::vector<aabb3f>();
+			predicted_node.getCollisionBoxes(nodedef_manager, cboxes, 0);
```
Comment (grorp) [inline src/client/game.cpp:3744]: As v-rob said, you don't need to allocate `cboxes` on the heap.

Instead of passing `0` for `neighbors`, you should pass the actual neighbors (`predicted_node.getNeighbors`) so that connected nodeboxes can be resolved correctly.

File: src/client/game.cpp [hunk 6]
```diff
@@ -3730,15 +3731,56 @@ bool Game::nodePlacement(const ItemDefinition &selected_def,
...
+
+			// Convert the node placement prediction position into BS space.
+			const v3f BS_space_p = v3f(
+				(f32)p.X * BS,
+				(f32)p.Y * BS,
+				(f32)p.Z * BS
+			);
```
Comment (grorp) [inline src/client/game.cpp:3751]: You can simply use `intToFloat(p, BS)`.

**Review by grorp — CHANGES_REQUESTED** [review]
(I haven't tested again yet.)

File: src/client/game.cpp (hunk 4 above)
Comment (rubenwardy) [reply src/client/game.cpp:3771]: You should never* use `new`, `delete`, `malloc`, or `free` in modern C++ - use smart pointers (std::unique_ptr) or stack allocation instead

(* - well, only if you're writing a low level data structure class that encapsulates these things)

**Comment by jordan4ibanez** [comment]
Due to an issue with your character's collision box Min Y becoming equal with the collision box Max Y of the thing you're standing on, I've had to duct tape a solution together with a note

File: src/client/game.cpp (hunk 4 above)
Comment (jordan4ibanez) [reply src/client/game.cpp]: Fixed in: 30e50ea2c4759dbdc59a93f98eb21239daf7e56c

File: src/client/game.cpp (hunk 5 above)
Comment (jordan4ibanez) [reply src/client/game.cpp]: >As v-rob said, you don't need to allocate cboxes on the heap.

Fixed in: 30e50ea2c4759dbdc59a93f98eb21239daf7e56c

File: src/client/game.cpp [hunk 7]
```diff
@@ -3730,15 +3731,53 @@ bool Game::nodePlacement(const ItemDefinition &selected_def,
...
+			predicted_node.getCollisionBoxes(nodedef_manager, &cboxes, predicted_node.getNeighbors(p, &map));
+
+			// Convert the node placement prediction position into BS space.
+			const v3f BS_space_p = intToFloat(p, BS);
+
+			// Create the player's collision box in BS real world space.
+			aabb3f cbox = player->getCollisionbox();
+			v3f pos = player->getPosition();
+			aabb3f player_cbox(cbox.MinEdge + pos,cbox.MaxEdge + pos);
+			for (uint i = 0; i < cboxes.size(); i++) {
```
Comment (grorp) [inline src/client/game.cpp:3753]: > I have no idea why this is failing on visual studio so if any core dev knows why please inform me

I think this is why the builds are failing: `uint` is no standardized type in C++.

Why don't you use a simple `for (auto node_cbox : cboxes)`?

Comment (jordan4ibanez) [reply src/client/game.cpp]: Now I am quite excited to see

Comment (jordan4ibanez) [reply src/client/game.cpp]: > > I have no idea why this is failing on visual studio so if any core dev knows why please inform me
> 
> I think this is why the builds are failing: `uint` is no standardized type in C++.
> 
> Why don't you use a simple `for (auto node_cbox : cboxes)`?

Fixed in: eab074b209425bb7f04fcbd549e846f6165d1f90

**Comment by grorp** [comment]
I'm not sure whether this PR can be merged as is. The timer reset you added fixes the "jump issue" if you hold RMB to place nodes, but if you only press RMB for a short time (which is what I usually do), it doesn't help. Placing nodes (without holding RMB) below yourself while jumping is still made way harder by this PR.

With MT's default initial jump velocity and gravity, the jump height is 1.0767 (considering that MT doubles the gravity for whatever reason). With this value, the difficulty is still acceptable.

However, in MCL2, the jump height is 1.0471 and this patch makes it nearly impossible to place a node (without holding RMB) below yourself during a jump. It seems that this was only possible before because the "don't build where you stand" check was broken. So at least MCL2 will likely have to adjust their initial jump velocity and gravity values if this is merged.

Maybe we could put the correct behavior behind a boolean physics override? Then games could first adjust their initial jump velocity and gravity values and enable the correct behavior afterwards.

**Comment by jordan4ibanez** [comment]
Calling all core devs, calling all core devs:
Please help grorp review my garbage.
Thank you.

**Comment by appgurueu** [comment]
> Looking for coredev opinions

This is tricky. I think we'll need to allow some kind of tolerance such that games can get behavior more similar to the previous behavior. Even though these games only ever "worked" in the first place due to broken behavior, we now can't break these games.

So I propose an intersection "tolerance". For each collision box, we can calculate the axis along which you "violate" the collision box "the least" (when jumping, this would be the Y axis). If this value is below the tolerance, we ignore it.

**Comment by jordan4ibanez** [comment]
If I shrink it in simulation, you might still fall through it

**Comment by appgurueu** [comment]
> If I shrink it in simulation, you might still fall through it

Indeed, and that's intended. The current behavior is broken, but by properly fixing it, we will break mods and games relying on the current broken-ness.

So I believe some kind of "tolerance" here can make for a good compromise that'll work well 90% of the time.

Note especially that games that don't want this tolerance at all (and instead want to increase player jump strength) can just set it to zero.

Note also that our collision handling code itself has some kind of tolerance where it pushes objects out of boxes they collide with [citation needed]. If these tolerances were made to match by default, it might work especially well.

**Comment by Zughy** [comment]
So what's left to do here? Unsubscribing

**Comment by grorp** [comment]
> So what's left to do here? Unsubscribing

As seen above, this still needs backwards compatibility, e.g. via a physics override and/or a "tolerance" value.

**Comment by jordan4ibanez** [comment]
No idea how to do that as I haven't touched this engine in months but I am open to core devs adding that in

**Comment by duncanacnud** [comment]
> This is tricky. I think we'll need to allow some kind of tolerance such that games can get behavior more similar to the previous behavior. Even though these games only ever "worked" in the first place due to broken behavior, we now can't break these games.

Maybe instead of some sort of tolerance, it could use the player's stepheight, and if the player is colliding "stepheight" into the block, it places the block and snaps the player to the top of it

**Comment by appgurueu** [comment]
That might work but I'd generally prefer not to "overload" something like stepheight that serves a specific purpose (which is not clipping recovery). Modders that want those things to be equal should just said that.

It could be argued that stepheight might provide some sensible tolerances "out of the box" even for mods that were written before such a change but then again, ~0.5 (the default in MTG-like games which lets you climb stairs and slabs) seems like it would be larger than what many games will want?

TL;DR: new physics override or such is probably still the best idea (and also not hard to do at all)
